src/daemon/global_rpc_protocol.rs

```diff
@@ -0,0 +1,96 @@
+use std::time::{Duration, Instant};
+
+use async_trait::async_trait;
+use earendil_crypt::Fingerprint;
+use earendil_packet::Dock;
+use futures_util::{future, FutureExt};
+use nanorpc::{nanorpc_derive, JrpcRequest, JrpcResponse, RpcTransport};
+use smol::Timer;
+
+use super::{
+    socket::{Endpoint, Socket},
+    DaemonContext,
+};
+
+pub const GLOBAL_RPC_DOCK: Dock = 100001;
+
+#[nanorpc_derive]
+#[async_trait]
+pub trait GlobalRpcProtocol {
+    async fn ping(&self, i: u64) -> u64;
+}
+
+pub struct GlobalRpcImpl {
+    ctx: DaemonContext,
+}
+
+impl GlobalRpcImpl {
+    pub fn new(ctx: DaemonContext) -> GlobalRpcImpl {
+        GlobalRpcImpl { ctx }
+    }
+}
+
+#[async_trait]
+impl GlobalRpcProtocol for GlobalRpcImpl {
+    async fn ping(&self, i: u64) -> u64 {
+        i
+    }
+}
+
+pub struct GlobalRpcTransport {
+    ctx: DaemonContext,
+    dest: Fingerprint,
+}
+
+impl GlobalRpcTransport {
+    pub fn new(ctx: DaemonContext, dest: Fingerprint) -> GlobalRpcTransport {
+        GlobalRpcTransport { ctx, dest }
+    }
+}
+
+#[async_trait]
+impl RpcTransport for GlobalRpcTransport {
+    type Error = anyhow::Error;
+
+    async fn call_raw(&self, req: JrpcRequest) -> Result<JrpcResponse, Self::Error> {
+        let endpoint = Endpoint::new(self.dest, GLOBAL_RPC_DOCK);
+        let socket = Socket::bind(self.ctx.clone(), None, None);
+        let mut retries = 0;
+        let max_retries = 3;
+        let mut timeout: Duration;
+
+        while retries <= max_retries {
```

Comment:
infinitely looping is probably better. the user can decide how long to wait by composing with `.timeout()` or similar.